modules/generator/processor/spanmetrics/spanmetrics.go

```diff
@@ -0,0 +1,128 @@
+package spanmetrics
+
+import (
+	"context"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+
+	gen "github.com/grafana/tempo/modules/generator/processor"
+	processor_util "github.com/grafana/tempo/modules/generator/processor/util"
+	"github.com/grafana/tempo/pkg/tempopb"
+	v1_trace "github.com/grafana/tempo/pkg/tempopb/trace/v1"
+	tempo_util "github.com/grafana/tempo/pkg/util"
+)
+
+type processor struct {
+	cfg Config
+
+	spanMetricsCallsTotal      *prometheus.CounterVec
+	spanMetricsDurationSeconds *prometheus.HistogramVec
+
+	// for testing
+	now func() time.Time
+}
+
+func New(cfg Config, tenant string) gen.Processor {
+	return &processor{
+		cfg: cfg,
+		now: time.Now,
+	}
+}
+
+func (p *processor) Name() string { return Name }
+
+func (p *processor) RegisterMetrics(reg prometheus.Registerer) error {
+	labelNames := []string{"service", "span_name", "span_kind", "span_status"}
```

Comment:
here: `span_status` is going to be an integer with values 0/1/2. That might be slightly confusing? do we want to turn this into an `error=true/false` at some point?

Reply:
We call `.String()` on `StatusCode` and the generated proto code uses an enum lookup to get the string label. So we are adding label values like `STATUS_CODE_UNSET` here.

- https://github.com/grafana/tempo/blob/metrics-generator-dev/modules/generator/processor/spanmetrics/spanmetrics.go#L110
- https://github.com/grafana/tempo/blob/metrics-generator-dev/pkg/tempopb/trace/v1/trace.pb.go#L177-L179

Example of the metrics that are generated:

![Screenshot 2022-02-18 at 15 05 36](https://user-images.githubusercontent.com/7748404/154697298-84e37ead-47e9-440b-895f-3d2f9dfd9698.png)

---

Regarding using `error=true/false` vs `span_status=STATUS_CODE_...`: I'm not sure. We already have this mix in various places:
- in Tempo search we allow searching on `error=true`, even though `error` is not necessarily a real attribute
- Grafana also shows `error=true` in the trace view because the trace view is based upon Jaeger

This processor is based upon [the spanmetrics processor from OTel Collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/spanmetricsprocessor) which exclusively works on OTLP. So that's why we only added the `span_status` label.

So for now I'd say we keep it simple and only add `span_status` (which aligns most closely to the underlying data) and if this is confusing/annoying we can look into adding an `error` label to the metrics.

Reply:
Ok, sounds good. 👍 

Thanks for the explanation!